### PR TITLE
test: deserialize `TransferState` and `Action` from json files

### LINKED ISSUE #51: test: deserialize `TransferState`/`Action` from json files
### Description
As introduced in #50 :
The test framework needs a folder which will contain all the json files representing both Action's (going from frontend to backend) as well as TransferState's. These json files will be used by both the frontend and backend for their respective testing. Using the same exact test json files for both frontend and backend ensures that both sides can trust that the intended app state updates is in sync.

The current tests are located in  `src-tauri/tests/tests`. All the instances of `TransferState` and `Action` should be deserialized from json files.

Location of the new folder:
* `src-tauri/tests/fixtures`

With subfolders:
*  `src-tauri/tests/fixtures/actions`
*  `src-tauri/tests/fixtures/states`

### Motivation
Better structured testing environment

### Resources
deserialize using serde_json:

* https://docs.rs/serde_json/latest/serde_json/

tracking issue: #50 

### To-do List

- [ ]

### PR DESCRIPTION
**Pull request by Oran-Dan**
# Description of change
I have added json files for the Transferstates and Actions and changed the test files to work with them.
Some info in the json files is now hardcoded, I kept the previously accompanying code in the test files commented out but not deleted.

## Links to any relevant issues
fixes issue #51 .

## How the change has been tested
I have tested the changes by running again the same test files I edited.
I have also added a 'async fn pretty_print_test' for testing the serialization and deserialization json output.

## Definition of Done checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src-tauri/.snapshot
Comment (daniel-mader) [inline src-tauri/.snapshot:1]: What is this file needed for?

Comment (Oran-Dan) [reply src-tauri/.snapshot:1]: Well it seems to be an auto-generated file created by either some "vitest" node module in the package file or some dependencies in the cargo.toml file eitherway I couldnt find any specific info about it other than it seems to originate from or own code, not any vscode extensions. I'm guessing we don't need it, but yeah I forgot to ask and check about this specific file. Should I delete it?

Comment (nanderstabel) [reply src-tauri/.snapshot:1]: The `.snapshot` file is the Stronghold binary used for storing credentials in some of the tests. On successful runs of the tests it should be automatically removed, but when a test fails, it may also fail to remove it again. 

So indeed you can remove it and add:
```
.snapshot

```

(including newline) to the end of the `src-tauri/.gitignore` file so it won't be added anymore in the future.

File: src-tauri/tests/tests/fixtures-get_state/actions/create_new1.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "type": "[App] Get state"
+}
```
Comment (nanderstabel) [inline src-tauri/tests/tests/fixtures-get_state/actions/create_new1.json]: For gits version control it is preferred to have a newline at the end of text files. So you can add a newline to each of your json files!

More info in case you're interested: https://medium.com/@alexey.inkin/how-to-force-newline-at-end-of-files-and-why-you-should-do-it-fdf76d1d090e

File: src-tauri/tests/tests/fixtures-get_state/states/unlock_storage-app_state.json
Comment (nanderstabel) [inline src-tauri/tests/tests/fixtures-get_state/states/unlock_storage-app_state.json:1]: This file is not being used in the tests. Can it be removed?

File: src-tauri/tests/tests/qr_code_scanned.rs [hunk 2]
```diff
@@ -23,6 +25,55 @@ use oid4vci::{
...
+    
+
+    let s = serde_json::to_string_pretty(&test).unwrap();
+    println!("{}", s);
+    let ds = serde_json::from_str::<TransferState>(&s).unwrap();
+    println!("{:#?}", ds);
+    let dss = json_example::<TransferState>("tests/tests/fixtures-qr_code_scanned/states/handle_oid4vp_2.json");
+    println!("{:#?}", dss);
+}
+
```
Comment (nanderstabel) [inline src-tauri/tests/tests/qr_code_scanned.rs]: This can be removed right?

File: src-tauri/tests/tests/fixtures-load_dev_profile/states/load_dev_profile.json [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            }
+          },
+          "metadata": {
+            "is_favorite": false,
+            "display": {
+              "icon": null,
+              "color": null,
+              "name": null
+            }
+          }
```
Comment (nanderstabel) [inline src-tauri/tests/tests/fixtures-load_dev_profile/states/load_dev_profile.json]: Since the the values in metadata are all default values I think we should just omit them, same for metadata in line 69.

That means however that the `#[serde(default)]` attribute should be added above the`metadata` field in the `DisplayCredential` struct in `verifiable_credential_record.rs`.

File: src-tauri/tests/tests/load_dev_profile.rs [hunk 4]
```diff
@@ -13,35 +12,22 @@ async fn test_load_dev_profile() {
     setup_state_file();
     setup_stronghold();
 
-    let personal_information = PERSONAL_INFORMATION.clone().display_credential;
-    let drivers_license_credential = DRIVERS_LICENSE_CREDENTIAL.clone().display_credential;
+    // let personal_information = PERSONAL_INFORMATION.clone().display_credential;
+    // let drivers_license_credential = DRIVERS_LICENSE_CREDENTIAL.clone().display_credential;
```
Comment (nanderstabel) [inline src-tauri/tests/tests/load_dev_profile.rs]: All the commented out code in this file can be removed!

File: src-tauri/tests/tests/fixtures-load_dev_profile/states/load_dev_profile.json (hunk 3 above)
Comment (nanderstabel) [reply src-tauri/tests/tests/fixtures-load_dev_profile/states/load_dev_profile.json]: Same goes for `metadata` in `handle_oid4vp_1.json` and `handle_oid4vp_2.json`.

File: src-tauri/tests/tests/qr_code_scanned.rs [hunk 5]
```diff
@@ -45,8 +96,11 @@ async fn test_qr_code_scanned_read_credential_offer() {
             .unwrap(),
     });
 
-    let credential_issuer: url::Url = "http://192.168.1.127:9090".parse().unwrap();
+//    let credential_issuer: url::Url = "http://192.168.1.127:9090".parse().unwrap();
```
Comment (nanderstabel) [inline src-tauri/tests/tests/qr_code_scanned.rs]: Can be removed

File: src-tauri/tests/tests/qr_code_scanned.rs [hunk 6]
```diff
@@ -178,7 +175,7 @@ async fn test_qr_code_scanned_handle_oid4vp_authorization_request() {
 
     let credentials = vec![verifiable_credential_record.display_credential.clone()];
 
-    let uuid = verifiable_credential_record.display_credential.id.clone();
+//    let uuid = verifiable_credential_record.display_credential.id.clone();
```
Comment (nanderstabel) [inline src-tauri/tests/tests/qr_code_scanned.rs]: Can be removed

**Comment by nanderstabel** [comment]
The intended structure for the fixture files is as follows:
Location of the new folder:

    src-tauri/tests/fixtures

With subfolders:

    src-tauri/tests/fixtures/actions
    src-tauri/tests/fixtures/states

In your current update they are structured like this:
```
src-tauri/tests/tests/fixtures-<test-name>/actions
src-tauri/tests/tests/fixtures-<test-name>/states
```

I think it is better to throw all state json files into `src-tauri/tests/fixtures/states` since in the future 'mermaid' update all the `<test-name>`'s will not exist anymore. That also means though that the names of the state json files must be a bit more explicit so that it is easier to understand what's inside. For example the current `tests/tests/fixtures-qr_code_scanned/states/read_cred.json` should be renamed to something like `tests/fixtures/states/active_profile_with_credential_offer.json`.

The Actions are easier to rename, for example the current `tests/tests/fixtures-get_state/actions/create_new1.json` should be renamed to `tests/fixtures/actions/app_get_state.json`, named after the actual `ActionType`. 
Some are a bit trickier, like `tests/tests/fixtures-qr_code_scanned/actions/handle_oid4vp_1.json` and `tests/tests/fixtures-qr_code_scanned/actions/handle_siopv2_1.json`. They can be renamed to `tests/fixtures/actions/qr_code_scanned_with_vp_token.json` and `tests/fixtures/actions/qr_code_scanned_with_id_token.json` respectfully.

Other than that some of the code does not seem to be formatted as intended. You can easily formatted by using `cargo fmt`. 

If these minor issues are solved than this PR is good to go!

File: src-tauri/src/verifiable_credential_record.rs [hunk 7]
```diff
@@ -1,3 +1,5 @@
+use std::default;
+
```
Comment (nanderstabel) [inline src-tauri/src/verifiable_credential_record.rs]: This unused import seem to have slipped in!

File: src-tauri/src/state/user_prompt.rs [hunk 8]
```diff
@@ -69,11 +76,11 @@ pub struct ShareCredentials {
 #[serde(untagged)]
 #[ts(export, export_to = "bindings/user-prompt/CurrentUserPrompt.ts")]
 pub enum CurrentUserPrompt {
+    PasswordRequired(PasswordRequired),
```
Comment (nanderstabel) [inline src-tauri/src/state/user_prompt.rs]: Minor issue: If I recall correctly moving this up was initially in order to temporarily solve the weird deserialization issue. Since that has been resolved now with the help of `#[serde(deny_unknown_fields)]` `PasswordRequired(PasswordRequired),` can be placed back to it's original position below `    CredentialOffer(CredentialOffer),`.

In case if there's another reason for changing the order in this enum, then please also commit the corresponding change that `ts_rs` will generate in `bindings/user-prompt/CurrentUserPrompt.ts` when you execute `cargo test` in `src-tauri`.

**Comment by nanderstabel** [comment]
Thanks for resolving everything! I just commented two more small issues. Also the `fixtures` folder is still located in `src-tauri/tests/tests` while a better location imo would be to move it up one level to `src-tauri/tests`

**Review by nanderstabel — DISMISSED** [review]
Looks good to me, great work! :partying_face:

File: src-tauri/tests/fixtures/states/active_pf_password_required.json
Comment (daniel-mader) [inline src-tauri/tests/fixtures/states/active_pf_password_required.json:1]: What does `pf` stand for? Profile? Why abbreviate?

File: src-tauri/tests/fixtures/states/pf_with_driverl_and_personal_info_redirect_me.json
Comment (daniel-mader) [inline src-tauri/tests/fixtures/states/pf_with_driverl_and_personal_info_redirect_me.json:1]: I suggest renaming the file to something like `profile_with_two_credentials_prompt_redirect_me` makes it even more readable.

**Review by daniel-mader — COMMENTED** [review]
Nice work! 🎉 Please consider renaming some json files to more descriptive names.

**Comment by nanderstabel** [comment]
> Nice work! 🎉 Please consider renaming some json files to more descriptive names.

Naming can be quite tricky sometimes in programming, and this is no exception. So if we want to do it right(-ish) from the start I would propose these name changes: 

```
active_pf_password_required.json                    -> profile_password_required.json
active_pf_redirect_me.json                          -> profile_redirect_me.json
no_profile_redirect_welcome.json                    -> no_profile_redirect_welcome.json
pf_accept_connect.json                              -> profile_accept_connection.json
pf_credential_offer.json                            -> profile_credential_offer.json
pf_with_did_redirect_me.json                        -> profile_did_redirect_me.json
pf_with_driverl_and_personal_info_redirect_me.json  -> profile_with_two_credentials_redirect_me.json
pf_with_person_info_redirect_me.json                -> profile_with_credential_redirect_me.json
pf_with_person_info_share_cred.json                 -> profile_with_credential_share_credentials.json
```

Alternatively I would perhaps even prefer to omit the `profile_` prefix since most state json files will have an `active_profile` and thus it does not add any descriptive benefit.

**Review by daniel-mader — APPROVED** [review]
